MaskRCNN/model/generalized_rcnn.py

```diff
@@ -312,15 +321,38 @@ def roi_heads(self, image, features, proposal_boxes, targets, inputs):
                                                                        image_shape2d[i],
                                                                        pad_border=False,
                                                                        verbose_batch_index=i)
-                    per_image_fg_labels.append(single_image_fg_labels)
-                    per_image_target_masks_for_fg.append(single_image_target_masks_for_fg)
-
-                target_masks_for_fg = tf.concat(per_image_target_masks_for_fg, axis=0) # Num_fg_boxes x 1 x (H_roi_mask*2) x (W_roi_mask*2)
-
-                proposal_fg_labels = tf.concat(per_image_fg_labels, axis=0) # 1-D Num_fg_boxes
-
-                target_masks_for_fg = tf.squeeze(target_masks_for_fg, 1, 'sampled_fg_mask_targets') # Num_fg_boxes x (H_roi_mask*2) x (W_roi_mask*2)
-                mask_loss = maskrcnn_loss(mask_logits, proposal_fg_labels, target_masks_for_fg)
+                    if roi_head_per_image:
+                        single_image_roi_feature_maskrcnn = tf.gather(roi_feature_maskrcnn, single_image_fg_indices)
+                        single_image_mask_logits = maskrcnn_head_func('maskrcnn',
```

Comment:
I would guess that there's a performance penalty for doing this. I think we should calculate mask_logits on the batch. You can always do `single_image_mask_logits = mask_logits[i]`